tests/handler/handler_copy.cpp

```diff
@@ -17,334 +22,887 @@ using mode_t = cl::sycl::access::mode;
-  static cl::sycl::accessor<dataT, dims, mode, target> get(
-      cl::sycl::handler& cgh, cl::sycl::buffer<dataT, dims>& buf) {
-    return buf.template get_access<mode, target>(cgh);
+
+template <typename dataT, int dims, typename initOp>
+class fill_kernel;
+
+template <typename dataT, int dims, template <typename, int> class initOp>
+void fill_buffer(cl::sycl::queue& queue, cl::sycl::buffer<dataT, dims>& buf,
+                 initOp<dataT, dims> init_op) {
+  queue.submit([&](cl::sycl::handler& cgh) {
+    auto acc = buf.template get_access<mode_t::discard_write>(cgh);
+    cgh.parallel_for<fill_kernel<dataT, dims, initOp<dataT, dims>>>(
+        buf.get_range(), [=](cl::sycl::id<dims> id) { acc[id] = init_op(id); });
+  });
+}
+
+template <template <int> class T, int dims, size_t default_value>
+struct range_id_helper {};
+
+template <template <int> class T, size_t default_value>
+struct range_id_helper<T, 1, default_value> {
+  static T<1> make(size_t d0, size_t, size_t) { return T<1>{d0}; }
+
+  template <template <int> class other, int other_dims,
+            size_t dv = default_value>
+  static T<1> cast(const other<other_dims>& o) {
+    return T<1>{o[0]};
   }
 };
-template <typename dataT, int dims, mode_t mode>
-struct accessor_helper<dataT, dims, mode, target_t::host_buffer> {
-  static constexpr target_t target = target_t::host_buffer;
-  static cl::sycl::accessor<dataT, dims, mode, target> get(
-      cl::sycl::handler& cgh, cl::sycl::buffer<dataT, dims>& buf) {
-    return buf.template get_access<mode>();
+
+template <template <int dims> class T, size_t default_value>
+struct range_id_helper<T, 2, default_value> {
+  static T<2> make(size_t d0, size_t d1, size_t) { return T<2>{d0, d1}; }
+
+  template <template <int> class other, int other_dims,
+            size_t dv = default_value>
+  static T<2> cast(const other<other_dims>& o) {
+    return T<2>{o[0], other_dims >= 2 ? o[1] : dv};
   }
 };
-template <typename dataT, int dims, mode_t mode>
-struct accessor_helper<dataT, dims, mode, target_t::local> {
-  static constexpr target_t target = target_t::local;
-  static cl::sycl::accessor<dataT, dims, mode, target> get(
-      cl::sycl::handler& cgh, cl::sycl::buffer<dataT, dims>& buf) {
-    const auto elemsPerDim =
-        test_single_copy_function_base<dataT, dims, dims>::numElems / dims;
-    auto allocationSize =
-        buffer_helper<dataT, dims>::construct_range(elemsPerDim);
-    return cl::sycl::accessor<dataT, dims, mode, target>(allocationSize, cgh);
+
+template <template <int dims> class T, size_t default_value>
+struct range_id_helper<T, 3, default_value> {
+  static T<3> make(size_t d0, size_t d1, size_t d2) { return T<3>{d0, d1, d2}; }
+
+  template <template <int> class other, int other_dims,
+            size_t dv = default_value>
+  static T<3> cast(const other<other_dims>& o) {
+    return T<3>{o[0], other_dims >= 2 ? o[1] : dv, other_dims == 3 ? o[2] : dv};
```

Comment:
Naming a class "helper" does not really help the reader. :-)
Add a few comments to explain the high-level behavior.

Reply:
Well, I just followed the conventions of the previous implementation ;-). Sure, will do.

Reply:
Yes, I know... Thank you for making the world better and more understandable. :-)